### ISSUE TITLE
Report bug

### ISSUE THREAD

**Issue by Alxaline** [jcreinhold/intensity-normalization#3]
Hello,

First of all, I congratulate you for this nice tool.

I have a few comments to make : 
In the readocs you indicate the command line with an "_" but it's an "-"
ex : zscore_normalize -> zscore-normalize

For the hm method if i use my image not skull-stripped with a mask and my image skull-stripped without mask, i have some difference.

Also, it will be nice, if we can disable the log for the count =)

For me the Ravel method don't work, it run for 2 of my image and then, i got this error :

### **### 2018-10-09 11:07:06,506 - intensity_normalization.exec.ravel_normalize - ERROR - expected str, bytes or os.PathLike object, not NoneType
Traceback (most recent call last):
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/ravel_normalize.py", line 77, in main
    mask_fns = io.glob_nii(args.mask_dir)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/utilities/io.py", line 48, in glob_nii
    fns = sorted(glob(os.path.join(dir, '*.nii*')))
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
(intensity_normalization) h@h-AERO-15XV8:/data/PythonProjects/standardization$ ravel-normalize -i /media/h/work_data/Patient_data_ST_ANNE/NIFTI_normalization/1_5T_FLAIR/img -o /media/h/work_data/Patient_data_ST_ANNE/NIFTI_normalization/1_5T_FLAIR/ravel -m /media/h/work_data/Patient_data_ST_ANNE/NIFTI_normalization/1_5T_FLAIR/mask -c flair -v -p
2018-10-09 11:07:54,286 - intensity_normalization.exec.ravel_normalize - INFO - Normalizing the images according to RAVEL
2018-10-09 11:07:54,648 - intensity_normalization.normalize.ravel - INFO - Applying WhiteStripe to image XX_1_5T_FLAIR_1mm_register (1/22)
2018-10-09 11:07:55,599 - intensity_normalization.normalize.ravel - INFO - Creating control mask for image XX_1_5T_FLAIR_1mm_register (1/22)
2018-10-09 11:09:06,540 - intensity_normalization.normalize.ravel - INFO - Applying WhiteStripe to image YY_1_5T_FLAIR_1mm_register (2/22)
2018-10-09 11:09:09,474 - intensity_normalization.normalize.ravel - INFO - Starting registration for image YY_1_5T_FLAIR_1mm_register (2/22)
2018-10-09 11:10:00,504 - intensity_normalization.normalize.ravel - INFO - Creating control mask for image YY_1_5T_FLAIR_1mm_register (2/22)
/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/segmentation/atropos.py:141: UserWarning:

ERROR: Non-zero exit status!

2018-10-09 11:10:04,841 - intensity_normalization.exec.ravel_normalize - ERROR - list index out of range
Traceback (most recent call last):
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/ravel_normalize.py", line 86, in main
    use_fcm=args.use_fcm)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/normalize/ravel.py", line 90, in ravel_normalize
    use_fcm=use_fcm)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/normalize/ravel.py", line 204, in image_matrix
    mrf=smoothness, use_fcm=use_fcm))
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/utilities/csf.py", line 52, in csf_mask
    res = img.kmeans_segmentation(3, kmask=brain_mask, mrf=mrf)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/segmentation/kmeans.py", line 47, in kmeans_segmentation
    kmimage = atropos(a = kmimage, m = mrf, c = '[5,0]', i = 'kmeans[%s]'%(str(k)), x = kmask)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/segmentation/atropos.py", line 144, in atropos
    probimgs = [iio2.image_read(probsout[0])]
IndexError: list index out of range**

Best regards,

**Comment by jcreinhold** [comment]
Hello @Alxaline, thank you very much for the feedback. I'll address the feedback point-by-point:

* I corrected the documentation issue, thank you for pointing that out.
* I implemented a fix in the RAVEL script because I encountered the same issue. Please use the `--use-fcm` flag in the `ravel-normalize` executable. The problem that you and I encountered lies in the *atropos* tissue segmentation method that is called through the ANTsPy program. I'll have to raise an issue with them about this problem, but in the meantime the program should work as expected with the `--use-fcm` flag enabled (run `ravel-normalize -h` for the exact call specifications).
* I need to investigate the problem with HM which I will get to by the end of this week. So standby for an update on that issue.

To get the updates please re-download the application in the v1.1-alpha pre-release. The issues you pointed out—excluding the HM problem—should be addressed in that pre-release. The HM normalization problem will be addressed in the full release, which again I should get to by Friday.

Thanks for the feedback and let me know if you still encounter any further problems.

**Comment by Alxaline** [comment]
Thank you for your quick and detailed answer.

A few new remarks : 
- For WhiteStripe and hm, it miss the add suffix "_norm" for the output files. And it will be nice, to add the method name in the suffix. 
- For plotting histograms with the whiteStripe and Zscore method, the intensity range starts at 0 and omits negative intensities values when no mask is used. This works when using a mask. In addition, I plot histograms with my own tools and it seems that using the WhiteStripe and Zscore method without a mask gives intensities peaks at low intensities. It seems to work perfectly with a mask.

Best regards,

**Comment by jcreinhold** [comment]
Update:

* I added the method name in the suffix, that is a good idea and I meant to do something for that before and never got around to it so thank you for prompting me
* I believe I fixed the histogram plotting routine. Previously, when the mask was not supplied I calculated the histogram for all intensities above zero. This failed for Z-score, WhiteStripe, and RAVEL which center some mean around zero. Now when no brain mask is provided the histogram of the whole image is plotted.
* The problem with WhiteStripe and Zscore *should* be fixed, previously I was estimating the foreground with the same method as above (everything above zero). The behavior now is to pick the intensities above the mean of the image which should exclude the majority of the background without taking away too much from the foreground. Please let me know if you continue to see odd behavior.

In your first comment, you stated that you observed a difference between the skull-stripped image and non-skull-stripped image when using the HM method? Could you provide a little bit more detail? 

I added some detail to the documentation for the HM method. Since it is doing piecewise linear histogram matching, I think if you provide a mask the image should also be skull-stripped. Otherwise you may learn an odd mapping voxels outside the brain mask may have greater intensity than those inside the image (e.g., fat). The upshot is that if you provide a brain mask, then your input image should be skull-stripped. So there should be a difference between the HM-normalized image with and without a mask.

Thanks for your feedback!

**Comment by Alxaline** [comment]
Hello,

I update my clone and now :  

- When i try to plot, i get this error : 
2018-10-15 10:40:41,527 - intensity_normalization.plot.hist - INFO - Creating histogram for image 1/22
2018-10-15 10:40:41,677 - intensity_normalization.exec.plot_hists - ERROR - /home/ncullen/Desktop/ANTsPy/itksource/Modules/IO/ImageBase/src/itkImageIOBase.cxx:370:
itk::ERROR: NiftiImageIO(0xdf292c0): Unknown component type: 0
Traceback (most recent call last):
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/plot_hists.py", line 57, in main
    ax = all_hists(args.img_dir, args.mask_dir, alpha=args.alpha, figsize=args.figsize, lw=args.linewidth)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/plot/hist.py", line 60, in all_hists
    mask = ants.image_read(mask_fn)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/core/ants_image_io.py", line 394, in image_read
    hinfo = image_header_info(filename)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/core/ants_image_io.py", line 320, in image_header_info
    retval = libfn(filename)
RuntimeError: /home/ncullen/Desktop/ANTsPy/itksource/Modules/IO/ImageBase/src/itkImageIOBase.cxx:370:

- For RAVEL, i try with --use-fcm and i get this one : 
2018-10-15 10:44:57,220 - intensity_normalization.exec.ravel_normalize - ERROR - expected str, bytes or os.PathLike object, not NoneType
Traceback (most recent call last):
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/ravel_normalize.py", line 77, in main
    mask_fns = io.glob_nii(args.mask_dir)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/utilities/io.py", line 48, in glob_nii
    fns = sorted(glob(os.path.join(dir, '*.nii*')))
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType

- It still don't add the suffix name 

- You have answered my question about the weird behaviour for HM with and without mask in your last comment. So Thank you for highlighting this point.

Best regards,

**Comment by jcreinhold** [comment]
* I have not encountered the first error before which states: `itk::ERROR: NiftiImageIO(0xdf292c0): Unknown component type: 0`. This is raised via ITK and is due to the image (what actually appears to be a mask for you) that is being used. What is the mask that you are trying to use with the `plot-hists` routine in this case? Does this error happen with every set of images you are trying to run `plot-hists` for? Can you provide the exact command you ran to create the error?

* The second error with RAVEL seems to be due to a problem with my documentation. Masks are *required* for RAVEL, not an option. I will update that in the documentation.

* If you do not see suffix names, I think you have not actually updated the software. How did you install the package? If you only ran `git pull` and you did not install the package with the `python setup.py develop` option, then you will not see the update. I'd run `pip uninstall intensity-normalization` the package and reinstall it (with the develop option since I'm pushing updates). You should get suffixes for all methods when the software is properly updated on your system. Note that suffixes are only added when saving images via the command line interface. Please let me know if you still encounter problems.

**Comment by Alxaline** [comment]
Hello,

I reinstall the environment and update software, so now suffix name works.

I still get an error with plot command. Here is the command line. Mask are the same that i use for normalization.
plot-hists -i /media/h/work_data/Patient_data/NIFTI_normalization_test/1_5T_FLAIR_3T_FLAIR/stripped  -m /media/h/work_data/Patient_data/NIFTI_normalization_test/1_5T_FLAIR_3T_FLAIR/mask -o hist_brut_1_5T_FLAIR_3T_FLAIR.png -v
2018-10-17 22:42:56,350 - intensity_normalization.plot.hist - INFO - Creating histogram for image 1/44
2018-10-17 22:42:56,493 - intensity_normalization.exec.plot_hists - ERROR - /home/ncullen/Desktop/ANTsPy/itksource/Modules/IO/ImageBase/src/itkImageIOBase.cxx:370:
itk::ERROR: NiftiImageIO(0x11df2a0): Unknown component type: 0
Traceback (most recent call last):
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/plot_hists.py", line 58, in main
    ax = all_hists(args.img_dir, args.mask_dir, alpha=args.alpha, figsize=args.figsize, lw=args.linewidth)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/plot/hist.py", line 63, in all_hists
    mask = ants.image_read(mask_fn)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/core/ants_image_io.py", line 394, in image_read
    hinfo = image_header_info(filename)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/core/ants_image_io.py", line 320, in image_header_info
    retval = libfn(filename)
RuntimeError: /home/ncullen/Desktop/ANTsPy/itksource/Modules/IO/ImageBase/src/itkImageIOBase.cxx:370:
itk::ERROR: NiftiImageIO(0x11df2a0): Unknown component type: 0

and for Ravel, the same error : 
ravel-normalize -i /media/h/work_data/Patient_data/NIFTI_normalization_test/1_5T_FLAIR_3T_FLAIR/stripped -o /media/h/work_data/Patient_data/NIFTI_normalization_test/1_5T_FLAIR_3T_FLAIR/ravel -m /media/h/work_data/Patient_data/NIFTI_normalization_test/1_5T_FLAIR_3T_FLAIR/mask -c flair --use-fcm -v -p
2018-10-17 22:50:02,752 - intensity_normalization.exec.ravel_normalize - INFO - Normalizing the images according to RAVEL
2018-10-17 22:50:02,933 - intensity_normalization.normalize.ravel - INFO - Applying WhiteStripe to image BEU1_1_5T_FLAIR_1mm_stripped_register (1/44)
2018-10-17 22:50:04,267 - intensity_normalization.normalize.ravel - INFO - Creating control mask for image BEU1_1_5T_FLAIR_1mm_stripped_register (1/44)
2018-10-17 22:50:16,503 - intensity_normalization.normalize.ravel - INFO - Applying WhiteStripe to image BEU1_3T_FLAIR_1mm_stripped_register (2/44)
2018-10-17 22:50:17,313 - intensity_normalization.exec.ravel_normalize - ERROR - /home/ncullen/Desktop/ANTsPy/itksource/Modules/IO/ImageBase/src/itkImageIOBase.cxx:370:
itk::ERROR: NiftiImageIO(0x12ef6d0): Unknown component type: 0
Traceback (most recent call last):
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/ravel_normalize.py", line 86, in main
    use_fcm=args.use_fcm)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/normalize/ravel.py", line 90, in ravel_normalize
    use_fcm=use_fcm)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/normalize/ravel.py", line 195, in image_matrix
    tmask = ants.image_read(masks[0])
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/core/ants_image_io.py", line 394, in image_read
    hinfo = image_header_info(filename)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/ants/core/ants_image_io.py", line 320, in image_header_info
    retval = libfn(filename)
RuntimeError: /home/ncullen/Desktop/ANTsPy/itksource/Modules/IO/ImageBase/src/itkImageIOBase.cxx:370:
itk::ERROR: NiftiImageIO(0x12ef6d0): Unknown component type: 0

Thank you for your replies.

Best regards,

**Comment by Alxaline** [comment]
I noticed that import ans didn't work so I compiled it manually, now i just get this error when i try to plot :+1: 
2018-10-18 00:28:05,259 - intensity_normalization.plot.hist - INFO - Creating histogram for image 1/44
2018-10-18 00:28:05,402 - intensity_normalization.exec.plot_hists - ERROR - 'long'
Traceback (most recent call last):
  File "/home/alexandre/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/plot_hists.py", line 58, in main
    ax = all_hists(args.img_dir, args.mask_dir, alpha=args.alpha, figsize=args.figsize, lw=args.linewidth)
  File "/home/alexandre/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/plot/hist.py", line 63, in all_hists
    mask = ants.image_read(mask_fn)
  File "/home/alexandre/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/ANTsPy/ants/core/ants_image_io.py", line 427, in image_read
    libfn = utils.get_lib_fn(_image_read_dict[pclass][ptype][ndim])
KeyError: 'long'

**Comment by Alxaline** [comment]
I got a question that i encounter with Ravel. For the normalization we obligatory need the same shape for all image ? 
 intensity_normalization.exec.ravel_normalize - ERROR - Cannot normalize because image p1_1mm_stripped_register needs to have same dimension as all other images ((250, 250, 160) != (240, 240, 163))
Traceback (most recent call last):
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/exec/ravel_normalize.py", line 86, in main
    use_fcm=args.use_fcm)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/normalize/ravel.py", line 90, in ravel_normalize
    use_fcm=use_fcm)
  File "/home/h/Applications/anaconda3/envs/intensity_normalization/lib/python3.6/site-packages/intensity_normalization-1.0.0-py3.6.egg/intensity_normalization/normalize/ravel.py", line 182, in image_matrix
    'as all other images ({} != {})'.format(base, img_data.shape, img_shape))
intensity_normalization.errors.NormalizationError: Cannot normalize because image DEB1_1_5T_FLAIR_1mm_stripped_register needs to have same dimension as all other images ((250, 250, 160) != (240, 240, 163))

**Comment by jcreinhold** [comment]
Hey there, yes ANTsPy doesn't have support for several data types. I generally cast the images to float32. FWIW, I have a script to cast an entire directory of NIfTI images to float32 which I uploaded here: [link](https://gist.github.com/jcreinhold/c49a83f84d13c941851991f95594259b).

Regarding your last point, yes, with RAVEL I unfortunately require that all images be the same size. This is due to the fact that all the images are co-registered to the first image in the dataset, and the method registration doesn't result in the registered image having the exact same dimensions as the target image. This is a problem because we use the brain mask from that first image to isolate the CSF in each image (see the paper for more details on why that is required). The only practical way that I could think to make the method work is to require that all images be the same size. There is for sure a better way to make this work, but I have not gotten around to implementing a  solution. I opened another issue regarding this: #4.

In the mean time, and for the limited utility that this provides, I do have a script to check the image dimensions in a directory, find the most common image size, and remove or move the non-standard images. The code for that script is here: [link](https://gist.github.com/jcreinhold/84480b94ab65607c14c7b9fa0e194f52).
